### PR TITLE
Picture Thumbnails support

### PR DESCRIPTION
**Pull request by monkeyWie**
**Description**
Add picture thumbnails support,use [https://github.com/nfnt/resize](https://github.com/nfnt/resize) lib.
effect:
image
image

:rotating_light: Before submitting your PR, please read [community](https://github.com/filebrowser/community), and indicate which issues (in any of the repos) are either fixed or closed by this PR. See [GitHub Help: Closing issues using keywords](https://help.github.com/articles/closing-issues-via-commit-messages/).

- [x] DO make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] DO make sure you are making a pull request against the **master branch** (left side). Also you should start *your branch* off *our master*.
- [x] DO make sure that File Browser can be successfully built. See [builds](https://github.com/filebrowser/community/blob/master/builds.md) and [development](https://github.com/filebrowser/community/blob/master/development.md).
- [x] DO make sure that related issues are opened in other repositories. I.e., the frontend, caddy plugins or the web page need to be updated accordingly.
- [x] AVOID breaking the continuous integration build.

:heart: Thank you!

### CODE REVIEW

**Comment by nambrosch** [comment]
@monkeyWie I noticed that nfnt/resize isn't updated anymore, there are some suggestions about alternatives here, have you checked them out?  https://github.com/nfnt/resize/issues/63

**Comment by monkeyWie** [comment]
@nambrosch oh,I'm not checked,I just saw [nfnt/resize#63](https://github.com/nfnt/resize/issues/63),maybe I can use `golang.org/x/image/draw` lib.

**Comment by o1egl** [comment]
A few years ago, I was working on an image processing service that was on the hot path. The winner among all tested libs was opencv compiled with libjpeg-turbo library. I don't think that performance will be an issue in that type of project. Therefore, I propose to use one of these libraries:

- https://github.com/muesli/smartcrop
- https://github.com/disintegration/imaging

File: http/preview.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		width = maxSize
+		height = 0
+	} else if height > maxSize && height > width {
+		width = 0
+		height = maxSize
+	} else {
+		width = 0
+		height = 0
+	}
+	m := resize.Resize(uint(width), uint(height), img, resize.Lanczos3)
```
Comment (o1egl) [inline http/preview.go:124]: Please use NearestNeighbor for faster processing.

File: http/preview.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	default:
+		return nil, errors.New("extension is not supported")
+	}
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+const maxSize = 1080
```
Comment (o1egl) [inline http/preview.go:104]: I don't think that 1080px is a good size for thumbnail

File: http/preview.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+func compressImage(decode func(r io.Reader) (image.Image, error), fd io.Reader) (*bytes.Buffer, image.Image, error) {
+	img, err := decode(fd)
+	if err != nil {
+		return nil, nil, err
+	}
+	buf := bytes.NewBuffer([]byte{})
```
Comment (o1egl) [inline http/preview.go:111]:
```suggestion
	buf := &bytes.Buffer{}
```

File: http/preview.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+func compressImageHandler(file *files.FileInfo, fd io.Reader) (*bytes.Buffer, error) {
+	var (
+		buf *bytes.Buffer
+		m   image.Image
+		err error
+	)
+
+	switch file.Extension {
+	case ".jpg", ".jpeg":
+		buf, m, err = compressImage(jpeg.Decode, fd)
```
Comment (o1egl) [inline http/preview.go:75]: It's not the right way to decode images. Just import decoders and use https://golang.org/pkg/image/#Decode function

File: http/http.go [hunk 5]
```diff
@@ -59,6 +59,7 @@ func NewHandler(store *storage.Storage, server *settings.Server) (http.Handler,
 	api.Handle("/settings", monkey(settingsPutHandler, "")).Methods("PUT")
 
 	api.PathPrefix("/raw").Handler(monkey(rawHandler, "/api/raw")).Methods("GET")
+	api.PathPrefix("/compress").Handler(monkey(compressHandler, "/api/compress")).Methods("GET")
```
Comment (o1egl) [inline http/http.go]: Let's call it /thumb or /thumbnail

**Review by o1egl — CHANGES_REQUESTED** [review]

File: http/preview.go (hunk 1 above)
Comment (monkeyWie) [reply http/preview.go:124]: > A few years ago, I was working on an image processing service that was on the hot path. The winner among all tested libs was opencv compiled with libjpeg-turbo library. I don't think that performance will be an issue in that type of project. Therefore, I propose to use one of these libraries:
> 
> * https://github.com/muesli/smartcrop
> * https://github.com/disintegration/imaging

Thanks for your advice,I haven't done image processing before,so I only found [nfnt/resize](https://github.com/nfnt/resize) from google,next I will try use these libraries.

**Comment by monkeyWie** [comment]
@o1egl Thanks for your advice,I haven't done image processing before,so I only found `nfnt/resize` from google,next I will try use these libraries.

File: http/preview.go (hunk 1 above)
Comment (monkeyWie) [reply http/preview.go:124]: ok.

File: http/preview.go (hunk 2 above)
Comment (monkeyWie) [reply http/preview.go:104]: Do you have a advice?

Comment (o1egl) [reply http/preview.go:104]: Let's match the size with the actual thumbnails size on the page.

**Pull request closed without merging**